flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/Segment.java

```diff
@@ -18,38 +18,80 @@
 
 package org.apache.flink.iteration.datacache.nonkeyed;
 
+import org.apache.flink.annotation.Internal;
 import org.apache.flink.core.fs.Path;
+import org.apache.flink.core.memory.MemorySegment;
 
-import java.io.Serializable;
+import java.util.List;
 import java.util.Objects;
 
-/** A segment represents a single file for the cache. */
-public class Segment implements Serializable {
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
 
+/** A segment contains the information about a cache unit. */
+@Internal
+public class Segment {
+
+    /** The path to the file containing cached records. */
     private final Path path;
 
-    /** The count of the records in the file. */
+    /** The count of the records in the segment. */
     private final int count;
 
-    /** The total length of file. */
-    private final long size;
+    /** The total length of file containing cached records. */
+    private long fsSize = -1L;
+
+    /** The memory segments containing cached records. */
+    private List<MemorySegment> cache;
+
+    Segment(Path path, int count, long fsSize) {
+        this.path = checkNotNull(path);
+        checkArgument(count > 0);
+        this.count = count;
+        checkArgument(fsSize > 0);
+        this.fsSize = fsSize;
+    }
 
-    public Segment(Path path, int count, long size) {
-        this.path = path;
+    Segment(Path path, int count, List<MemorySegment> cache) {
+        this.path = checkNotNull(path);
+        checkArgument(count > 0);
         this.count = count;
-        this.size = size;
+        this.cache = checkNotNull(cache);
+    }
+
+    void setCache(List<MemorySegment> cache) {
+        this.cache = checkNotNull(cache);
     }
 
-    public Path getPath() {
+    void setDiskInfo(long fsSize) {
+        checkArgument(fsSize > 0);
+        this.fsSize = fsSize;
+    }
+
+    boolean isOnDisk() {
+        return fsSize > 0;
+    }
+
+    boolean isCached() {
+        return cache != null;
+    }
+
+    Path getPath() {
         return path;
     }
 
-    public int getCount() {
+    int getCount() {
         return count;
     }
 
-    public long getSize() {
-        return size;
+    long getFsSize() {
+        checkState(fsSize > 0);
```

Comment:
Suppose the input datastream does not have any data, we should still be able to snapshot/persist this data stream onto disk, and be able to reload this datastream after restarting the Flink job, right?

Then we will need to support a segment with `fsSize = 0`.